packages/core/src/messagingProtocols/proteus/ProteusService/ProteusService.ts

```diff
@@ -173,4 +178,90 @@ export class ProteusService {
       state: response.errored ? PayloadBundleState.CANCELLED : PayloadBundleState.OUTGOING_SENT,
     };
   }
+
+  public async encrypt(
+    plainText: Uint8Array,
+    users: UserPreKeyBundleMap | UserClients,
+    domain: string = '',
+  ): Promise<{missing: UserClients; encrypted: OTRRecipients<Uint8Array>}> {
+    const {sessions, userClients} = await this.mapRecipantsToSessionsAndClients(users, domain, true);
+
+    const payload = await this.coreCryptoClient.proteusEncryptBatched(sessions, plainText);
+
+    return this.extractEncryptedAndMissingFromBatchedPayload(payload, userClients);
+  }
+
+  private async mapRecipantsToSessionsAndClients(
+    users: UserPreKeyBundleMap | UserClients,
+    domain: string = '',
+    createSession = false,
+  ) {
+    const userClients = filterUserClientsWithoutPreKey(users);
+    const sessions: string[] = [];
+
+    for (const userId in users) {
+      const clientIds = userClients[userId];
+
+      for (const clientId of clientIds) {
+        const sessionId = this.cryptographyService.constructSessionId(userId, clientId, domain);
+
+        if (createSession) {
+          const sessionExists = (await this.coreCryptoClient.proteusSessionExists(sessionId)) as unknown as boolean;
+
+          if (!sessionExists) {
+            const userQualifiedId = {id: userId, domain};
+            await this.createSession(userQualifiedId, clientId);
```

Comment:
hummm maybe we shouldn't create sessions one-by-one. We should probably implement batching session creation (I believe there is a backend endpoint to query multiple user/client prekeys at once)

Reply:
The endpoint is `POST users/list-prekeys`